### ISSUE TITLE
Use of java.util.Optional could lead to undesired dependency on Guava

### ISSUE THREAD

**Issue by dveijck** [google/FreeBuilder#153]
In case you have a project that has no dependency on Guava, FreeBuilder could generate uncompilable Builder code in case a `java.util.Optional` property is used that is not 'unboxable'.

E.g. consider the following class in a project that has **no** Guava on the classpath:
```java
package example;
...

@FreeBuilder
public interface Person {

    String getName();
    Optional<Integer> getAge();
    Optional<String> getFavoriteComicBook();

    class Builder extends Person_Builder {}
}
```

When compiling the code it results in the following `Person_Builder` (only relevant snippets are shown):

```java
package example;
...
import com.google.common.base.Preconditions; // <------- This should not be in the import statement as the class com.google.common.base.Preconditions is not on the classpath!
...

@Generated("org.inferred.freebuilder.processor.CodeGenerator")
abstract class Person_Builder {

  ...

  public Person.Builder setAge(int age) {
    this.age = age; // <------- Seems OK, as this is a primitive.
    return (Person.Builder) this;
  }

  public Person.Builder setAge(Optional<? extends Integer> age) {...}

  public Person.Builder setNullableAge(@Nullable Integer age) {...}

  ...

  public Person.Builder setFavoriteComicBook(String favoriteComicBook) {
    this.favoriteComicBook = Preconditions.checkNotNull(favoriteComicBook); // <------- This is uncompilable as the class com.google.common.base.Preconditions is not on the classpath!
    return (Person.Builder) this;
  }

  public Person.Builder setFavoriteComicBook(Optional<? extends String> favoriteComicBook) {...}

  public Person.Builder setNullableFavoriteComicBook(@Nullable String favoriteComicBook) {...}

  ...

}
```

As you can see the `Optional<Integer>` case is handled correctly (as `java.lang.Integer` is 'unboxable' into `int` primitive), whereas the `Optional<String>` case leads to undesired dependency on Guava (as `java.lang.String` is not 'unboxable').

The specific code seems to be generated by the method `org.inferred.freebuilder.processor.OptionalPropertyFactory#addSetter`, specifically in the following code:

```java
      ...
      if (unboxedType.isPresent()) {
        code.addLine("  this.%1$s = %1$s;", property.getName()); // <------- This is fine if an object is unboxable.
      } else {
        code.addLine("  this.%1$s = %2$s.checkNotNull(%1$s);",
            property.getName(), Preconditions.class); // <------- This should take into account the cases where no Guava is available.
      }
      ...
```

**Comment by chrisalice** [comment]
Thanks, @dveijck — that's a most excellent bug report! You should find this is fixed in [v1.10.2](https://github.com/google/FreeBuilder/releases/tag/v1.10.2).

**Comment by dveijck** [comment]
Hi @chrisalice, thanks for the very quick reponse. Unfortunately I couldn't verify the fix earlier, but I checked it this morning and the dependency on Guava is now gone 👍 .

Many thanks for the fix.

**Comment by dveijck** [comment]
Hi @chrisalice,

Seems I was cheering a little too early. I believe there is still an issue with a undesired *indirect* dependency on Guava in case a `java.util.Optional` is used.

The `example.Person` class in the issue description has multiple fields. This results in the toString() method being generated in the inner `example.Person_Builder.Value` class as follows:

```java
package example;

...

@Generated("org.inferred.freebuilder.processor.CodeGenerator")
abstract class Person_Builder {

  ...
  
  private static final class Value implements Person {
    ...
    
    @Override
    public String toString() {
      return "Person{"
          + COMMA_JOINER.join("name=" + name, (age != null ? "age=" + age : null)) // <------- Compilation fails here as the COMMA_JOINER static final field is not generated.
          + "}";
    }
    
    ...

  }
  ...

}
```

When I include the Guava dependency everything compiles fine, because the field COMMA_JOINER gets generated, but it is however of type `com.google.common.base.Joiner` which belongs to the Guava dependency.

I'm not sure as I didn't debug the annotation processing, but the bug seems to be in the `org.inferred.freebuilder.processor.CodeGenerator#addValueType`:

```java
...

public class CodeGenerator {

  ...

  private static void addConstantDeclarations(Metadata metadata, SourceBuilder body) {
    if (body.feature(GUAVA).isAvailable() && metadata.getProperties().size() > 1) {
      body.addLine("")
          .addLine("private static final %1$s COMMA_JOINER = %1$s.on(\", \").skipNulls();",
              Joiner.class); // <------- This looks good, the COMMA_JOINER static final field only gets generated in case Guava is available.
    }
  }

  ...

  private static void addValueType(SourceBuilder code, Metadata metadata) {
    ...

    // toString
    if (metadata.standardMethodUnderride(StandardMethod.TO_STRING) == ABSENT) {
      code.addLine("")
          .addLine("  @%s", Override.class)
          .addLine("  public %s toString() {", String.class)
          .add("    return \"%s{", metadata.getType().getSimpleName());
      switch (metadata.getProperties().size()) {
        case 0: {
          code.add("}\";\n");
          break;
        }

        case 1: {
          Property property = getOnlyElement(metadata.getProperties());
          if (property.getCodeGenerator().getType() == Type.OPTIONAL) {
            code.add("\" + (%1$s != null ? \"%1$s=\" + %1$s : \"\") + \"}\";\n",
                property.getName()); // <------- When there is only one property and it is optional, everything is fine as there is no need to join anything with the COMMA_JOINER.
          } else {
            ...
          }
          break;
        }

        default: {
          // If one or more of the properties are optional, use COMMA_JOINER for readability.
          // Otherwise, use string concatenation for performance.
          if (any(metadata.getProperties(), IS_OPTIONAL)) {
            code.add("\"\n")
                .add("        + COMMA_JOINER.join(\n"); // <------- It looks as if there is more than one property and one or more of them are optional, COMMA_JOINER is always used whether or not it was generated. I.e. there is no 'Guava-safe' check.
            ...
          } else {
            ...
          }
          break;
        }
      }
      code.addLine("  }");
    }

  ...

}
```

So in short, the way the `toString()` method is generated for the value class is not completely working out in case someone has no dependency on Guava.

Do you want me to create a new issue for this or do you feel this belongs to the same case?

Best regards

**Comment by chrisalice** [comment]
Thanks, @dveijck! Another superb diagnosis. I've opened #157 to track.
